### PR TITLE
ON HOLD. need to not have a PR that is 95 files...

### PR DESCRIPTION
**Pull request by cnishina**
- Leave node scripts/errorTest.js commented out since it has something
to do with the control flow where the exit code is not set. Also the
breakpointhook.js file has issues we will need to address (Error: Cannot find
module).
- Add TODO block for slowHttpAndTimeout tests.
- Update spec/errorTest to use const, let over var and async/await.
- Add SELENIUM_PROMISE_MANAGER flag to plugins tests (currently missing)
- Turn on the rest of test suite prior to updating to selenium 4
(basicConf using blocking proxy, unit tests, dependency tests, and
typing tests)

### CODE REVIEW

File: spec/errorTest/shardedFailureConf.js [hunk 1]
```diff
@@ -2,7 +2,8 @@ var env = require('../environment.js');
 
 exports.config = {
   seleniumAddress: env.seleniumAddress,
-
+  SELENIUM_PROMISE_MANAGER: false,
+
```
Comment (CrispusDH) [inline spec/errorTest/shardedFailureConf.js]: redundant empty line

File: spec/errorTest/slowHttpAndTimeoutConf.js [hunk 2]
```diff
@@ -2,7 +2,8 @@ var env = require('../environment.js');
 
 exports.config = {
   seleniumAddress: env.seleniumAddress,
-
+  SELENIUM_PROMISE_MANAGER: false,
+
```
Comment (CrispusDH) [inline spec/errorTest/slowHttpAndTimeoutConf.js]: redundant empty line, some configuration does not have it, some have

File: spec/errorTest/baseCase/error_spec.js [hunk 3]
```diff
@@ -1,6 +1,6 @@
-describe('finding an element that does not exist', function() {
-  it('should throw an error', function() {
-    browser.get('index.html');
+describe('finding an element that does not exist', () => {
```
Comment (juliemr) [inline spec/errorTest/baseCase/error_spec.js]: As far as I can tell this file is never used - can we just remove it?

Comment (cnishina) [reply spec/errorTest/baseCase/error_spec.js]: Deleted.

**Comment by cnishina** [comment]
Also these "approved" changes will need another lgtm.

File: circle.yml [hunk 4]
```diff
@@ -52,7 +52,7 @@ jobs:
           name: Selenium Start
           background: true
           command: |
-            ./node_modules/webdriver-manager/bin/webdriver-manager update
+            ./bin/webdriver-manager update
```
Comment (cnishina) [inline circle.yml]: This was to just call the correct webdriver-manager because we need the binaries for some tests.

File: gulpfile.js [hunk 5]
```diff
@@ -41,7 +41,7 @@ gulp.task('tslint', function() {
 });
 
 gulp.task('lint', function(done) {
-  runSequence('tslint', 'jshint', 'format:enforce', done);
+  runSequence('tslint', 'format:enforce', done);
```
Comment (cnishina) [inline gulpfile.js]: Most of these are clean up to remove jshint and an extra done statement

File: lib/debugger.ts [hunk 6]
```diff
@@ -6,13 +6,11 @@ import {ProtractorBrowser} from './browser';
 import {Locator} from './locators';
 import {Logger} from './logger';
 import {Ptor} from './ptor';
-import * as helper from './util';
 let breakpointHook = require('./breakpointhook.js');
```
Comment (cnishina) [inline lib/debugger.ts]: These are unused and deleting.

File: circle.yml (hunk 4 above)
Comment (cnishina) [inline circle.yml]: This was to just call the correct webdriver-manager in circle. We need these binaries for some tests.

File: gulpfile.js (hunk 5 above)
Comment (cnishina) [inline gulpfile.js]: Most of these here are to clean up jshint and remove a not used done statement.

File: lib/driverProviders/attachSession.ts [hunk 7]
```diff
@@ -38,7 +38,7 @@ export class AttachSession extends DriverProvider {
   getNewDriver(): WebDriver {
     const httpClient = new http.HttpClient(this.config_.seleniumAddress);
     const executor = new http.Executor(httpClient);
-    const newDriver = WebDriver.attachToSession(executor, this.config_.seleniumSessionId);
+    const newDriver = WebDriver.attachToSession(executor, this.config_.seleniumSessionId, null);
```
Comment (cnishina) [inline lib/driverProviders/attachSession.ts]: Setting null here will ensure that the control flow option is set to null.

File: lib/driverProviders/driverProvider.ts [hunk 8]
```diff
@@ -3,7 +3,7 @@
  *  It is responsible for setting up the account object, tearing
  *  it down, and setting up the driver correctly.
  */
-import {Builder, Session, WebDriver} from 'selenium-webdriver';
+import {Builder, WebDriver} from 'selenium-webdriver';
```
Comment (cnishina) [inline lib/driverProviders/driverProvider.ts]: Removed unused Session.

File: lib/index.ts [hunk 9]
```diff
@@ -1,7 +1,7 @@
 import {ElementHelper, ProtractorBrowser} from './browser';
 import {ElementArrayFinder, ElementFinder} from './element';
 import {ProtractorExpectedConditions} from './expectedConditions';
-import {Locator, ProtractorBy} from './locators';
+import {ProtractorBy} from './locators';
```
Comment (cnishina) [inline lib/index.ts]: Another unused import.

File: lib/launcher.ts [hunk 10]
```diff
@@ -3,7 +3,6 @@
  * input configuration and launching test runners.
  */
 import * as fs from 'fs';
-import * as q from 'q';
```
Comment (cnishina) [inline lib/launcher.ts]: Removed q and moved this to native Promises.

File: lib/launcher.ts [hunk 11]
```diff
@@ -108,197 +107,170 @@ let initFn = function(configFile: string, additionalConfig: Config) {
...
-            let protractorError = e as ProtractorError;
-            ProtractorError.log(logger, errorCode, protractorError.message, protractorError.stack);
-            process.exit(errorCode);
-          } else {
-            logger.error(e.message);
-            logger.error(e.stack);
-            process.exit(ProtractorError.CODE);
-          }
-        });
+  // process.on('uncaughtException', (exc: (Error|string)) => {
```
Comment (cnishina) [inline lib/launcher.ts]: Why is this commented out? Well I don't think this is ever called. The ErrorHandler is now parsing errors in ~line 242-ish.

File: lib/launcher.ts [hunk 12]
```diff
@@ -108,197 +107,170 @@ let initFn = function(configFile: string, additionalConfig: Config) {
...
-
-              taskResults_.reportSummary();
-              let exitCode = 0;
-              if (taskResults_.totalProcessFailures() > 0) {
-                exitCode = RUNNERS_FAILED_EXIT_CODE;
-              } else if (taskResults_.totalSpecFailures() > 0) {
-                exitCode = 1;
+              taskResults_.add(result);
+              task.done();
+              createNextTaskRunner();
```
Comment (cnishina) [inline lib/launcher.ts]: Not sure:...This was an odd ball that we call the same function and not resolve it? This seems to work but not sure why.

File: lib/plugins.ts [hunk 13]
```diff
@@ -1,19 +1,12 @@
-import * as q from 'q';
```
Comment (cnishina) [inline lib/plugins.ts]: Remove q and webdriver promises.

File: lib/ptor.ts [hunk 14]
```diff
@@ -12,10 +12,10 @@ import {ProtractorBy} from './locators';
 export class Ptor {
   // Variables tied to the global namespace.
   browser: ProtractorBrowser;
-  $ = function(search: string): ElementFinder {
+  $ = function(_: string): ElementFinder {
     return null;
   };
-  $$ = function(search: string): ElementArrayFinder {
+  $$ = function(_: string): ElementArrayFinder {
```
Comment (cnishina) [inline lib/ptor.ts]: Minor clean up stuff. No biggie

File: lib/browser.ts [hunk 15]
```diff
@@ -479,25 +477,12 @@ export class ProtractorBrowser extends AbstractExtendedWebDriver {
...
    *
    * @example
-   * // Running against global browser, with control flow enabled
-   * browser.get('page1');
-   * browser.restart();
-   * browser.get('page2'); // 'page2' gotten by restarted browser
-   *
-   * // Running against global browser, with control flow disabled
+   * // Running against global browser
    * await browser.get('page1');
```
Comment (cnishina) [inline lib/browser.ts]: Doc clean up. No more `restartSync` and no more control flow.

File: lib/browser.ts [hunk 16]
```diff
@@ -831,125 +793,97 @@ export class ProtractorBrowser extends AbstractExtendedWebDriver {
...
     };
 
-    return this.driver.controlFlow()
-        .execute(() => {
-          return wdpromise.when(null);
-        })
-        .then(() => {
-          if (this.bpClient) {
-            return this.driver.controlFlow().execute(() => {
-              return this.bpClient.setWaitEnabled(false);
```
Comment (cnishina) [inline lib/browser.ts]: No more control flow!

File: lib/browser.ts [hunk 17]
```diff
@@ -962,17 +896,14 @@ export class ProtractorBrowser extends AbstractExtendedWebDriver {
    *
    * @param {number=} opt_timeout Number of milliseconds to wait for Angular to start.
    */
-  refresh(opt_timeout?: number) {
-    if (this.ignoreSynchronization) {
-      return this.driver.navigate().refresh();
+  async refresh(opt_timeout?: number) {
+    if (!await this.waitForAngularEnabled()) {
```
Comment (cnishina) [inline lib/browser.ts]: `ignoreSynchronization` is problematic. It was previously noted as deprecated and I think we should remove this property.

File: scripts/test.js [hunk 18]
```diff
@@ -67,65 +67,66 @@ passingTests.forEach(function(passing_test) {
...
+    .expectExitCode(1)
+    .expectErrors([
+      {message: 'Expected true to be false'},
+      {message: 'from setup'},
+      {message: 'from postTest passing'},
+      {message: 'from postTest failing'},
+      {message: 'from teardown'}
+    ]);
+
+// TODO(selenium4): turn these on when we figure out the correct error message handling.
```
Comment (cnishina) [inline scripts/test.js]: We need to talk about these and the following tests commented out. These are UnhandledPromiseRejectionWarnings. I'm not sure where to fix this but maybe for the next PR.

**Comment by cnishina** [comment]
Closing this one too!

**Pull request closed without merging**